File: 0450-delete-node-in-a-bst/0450-delete-node-in-a-bst.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public int inorderSucc(TreeNode root){
        int min = root.val;
        
        while(root!=null){
            min = root.val;
            root = root.left;
        }
        return min;
    }
    
    public TreeNode deleteNode(TreeNode root, int key) {
        if(root==null) return root;
        
        if(key<root.val)
            root.left = deleteNode(root.left,key);
        
        else if(key>root.val)
            root.right = deleteNode(root.right,key);
        
        // delete
        else{
            
            // deleting node only
            if(root.left==null) return root.right;
            else if(root.right==null) return root.left;
            
            // replace root value with inorder successor value and delete it
            else {                
                root.val = inorderSucc(root.right);            // replace
                root.right = deleteNode(root.right,root.val);  // delete
            }
        }
        
        return root;
            
                    
    }
}

File: 0450-delete-node-in-a-bst/0450-delete-node-in-a-bst.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def inorderSucc(self,root):
        val = root.val
        while root:
            val = root.val
            root = root.left
        return val
    
    
    def deleteNode(self, root: Optional[TreeNode], key: int) -> Optional[TreeNode]:
        if not root: return root
        
        # less to left
        if(key<root.val): root.left = self.deleteNode(root.left,key)
        
        # more to right
        elif(key>root.val): root.right = self.deleteNode(root.right,key)
        
        # equal
        else:
        
            # no child
            # one child
            if not root.left: return root.right
            elif not root.right: return root.left            
            
            # two child
            else:
                root.val = self.inorderSucc(root.right)
                root.right = self.deleteNode(root.right,root.val)
        return root
        